engine/src/main/java/org/terasology/particles/updating/ParticleUpdaterImpl.java

```diff
@@ -107,9 +127,38 @@ public void update(final float delta) {
     }
 
     @Override
-    public void configureEmitter(final ParticleEmitterComponent emitter,
-                                 final BiMap<Class<Component>, AffectorFunction> registeredAffectorFunctions,
-                                 final BiMap<Class<Component>, GeneratorFunction> registeredGeneratorFunctions) {
+    public void initialize() {
+        ModuleEnvironment environment = moduleManager.getEnvironment();
+
+        for (Class<?> type : environment.getTypesAnnotatedWith(RegisterParticleSystemFunction.class)) {
+            RegisterParticleSystemFunction annotation = type.getAnnotation(RegisterParticleSystemFunction.class);
+
+            if (!ParticleSystemFunction.class.isAssignableFrom(type)) {
+                logger.error("Cannot register particle system function {}, must be a subclass of ParticleSystemFunction", type.getSimpleName());
+            } else {
+                try {
+                    ParticleSystemFunction function = (ParticleSystemFunction) type.newInstance();
```

Comment:
Is there any way to ensure that a specific class can only have two sub-classes? This code relies on the fact that a `ParticleSystemFunction` is either a `GeneratorFunction`  or a `AffectorFunction` ...

Reply:
I'm not quite sure I follow